File: util/bin/i18n.pl
#!/usr/bin/perl -w
#
# Leibniz Bioactives Cloud
# Copyright 2017 Leibniz-Institut f. Pflanzenbiochemie
# 
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
# 
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#==============================================================================
#
# Update the resource files for different languages. All keys must 
# be present in the english localization, otherwise they will get lost.
#
# Do not call this Perl script directly, use i18n.sh instead.
#
# Note: Currently, the resource strings can NOT contain equal signs "="
# 
#
use feature qw( fc );

open(EN, $ARGV[0]) or die "Could not open $ARGV[0]\n";
open(XX, $ARGV[1]) or die "Could not open $ARGV[1]\n";

while($l = <EN>) {
    chomp $l;
    next if($l =~ /^#/);
    @f = split /=/, $l;
    $key = shift @f;
    $en{$key} = join "=", @f; 
}

while($l = <XX>) {
    chomp $l;
    next if($l =~ /^#/);
    @f = split /=/, $l;
    $key = shift @f;
    $xx{$key} = join "=", @f;
#    $xx{$f[0]} = $f[1];
}

close(EN);
close(XX);
open(XX, ">$ARGV[1]") or die "Could not open $ARGV[1]\n";

print XX "# 00
# 01  Leibniz Bioactives Cloud
# 02  Copyright 2017 Leibniz-Institut f. Pflanzenbiochemie
# 03
# 04  Licensed under the Apache License, Version 2.0 (the \"License\");
# 05  you may not use this file except in compliance with the License.
# 06  You may obtain a copy of the License at
# 07
# 08     http://www.apache.org/licenses/LICENSE-2.0
# 09
# 10  Unless required by applicable law or agreed to in writing, software
# 11  distributed under the License is distributed on an \"AS IS\" BASIS,
# 12  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# 13  See the License for the specific language governing permissions and
# 14  limitations under the License.
# 15
# 16  This file may be modified by the i18n utility!
# 17\n";

foreach $i ( sort { (fc $a) cmp (fc $b)} keys %en )
{
    if(defined $xx{$i})
    {
        print XX "$i=$xx{$i}\n";
    } else {
        print XX "$i=TRANSLATE: $en{$i}\n";
    }
}
close(XX);

File: integration-tests/src/main/java/de/ipb_halle/pageobjects/util/I18n.java
/*
 * Cloud Resource & Information Management System (CRIMSy)
 * Copyright 2021 Leibniz-Institut f. Pflanzenbiochemie
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */
package de.ipb_halle.pageobjects.util;

import java.text.MessageFormat;
import java.text.ParseException;
import java.util.Locale;
import java.util.ResourceBundle;

/**
 * Utility class for coping with internationalization via resource bundles both
 * from JSF and from the CRIMSy UI project.
 * 
 * @author flange
 */
public class I18n {
    private I18n() {
    }

    /*
     * Resource bundle keys for the JSF messages. Add keys if necessary, the bundle
     * can be found in the myfaces-api package.
     */
    public static final String JSF_REQUIRED_VALIDATION_ERROR_KEY = "javax.faces.component.UIInput.REQUIRED";

    /*
     * resource bundle names
     */
    private static final String JSF_BUNDLE_BASENAME = "javax.faces.Messages";
    private static final String UI_BUNDLE_BASENAME = "de.ipb_halle.lbac.i18n.messages";

    /**
     * @param key    JSF resource bundle key
     * @param locale
     * @return the entry from the JSF resource bundle corresponding to the
     *         {@code key}
     */
    public static String getJSFMessage(String key, Locale locale) {
        return getMessageFromBundle(JSF_BUNDLE_BASENAME, key, locale);
    }

    /**
     * Checks a given test string against the entry for the given {@code key} in the
     * JSF resource bundle. It also accounts for format strings in the entry.
     * 
     * @param test   test string
     * @param key    JSF resource bundle key
     * @param locale
     * @return {@code false} if the test string does not match the entry
     */
    public static boolean isJSFMessage(String test, String key, Locale locale) {
        String pattern = getJSFMessage(key, locale);
        if (pattern != null) {
            return isParsableMessage(test, pattern);
        } else {
            return false;
        }
    }

    /**
     * @param key    CRIMSy UI resource bundle key
     * @param locale
     * @return the entry from the CRIMSy UI resource bundle corresponding to the
     *         {@code key}
     */
    public static String getUIMessage(String key, Locale locale) {
        return getMessageFromBundle(UI_BUNDLE_BASENAME, key, locale);
    }

    /**
     * Checks a given test string against the entry for the given {@code key} in the
     * CRIMSy UI resource bundle. It also accounts for format strings in the entry.
     * 
     * @param test   test string
     * @param key    CRIMSy UI resource bundle key
     * @param locale
     * @return {@code false} if the test string does not match the entry
     */
    public static boolean isUIMessage(String test, String key, Locale locale) {
        String pattern = getUIMessage(key, locale);
        if (pattern != null) {
            return isParsableMessage(test, pattern);
        } else {
            return false;
        }
    }

    private static String getMessageFromBundle(String baseName, String key, Locale locale) {
        ResourceBundle bundle = ResourceBundle.getBundle(baseName, locale);
        if ((key != null) && (bundle.containsKey(key))) {
            return bundle.getString(key);
        } else {
            String error = String.format("Could not locate message key '%s' in bundle '%s' for locale %s.", key,
                    baseName, locale.toString());
            throw new RuntimeException(error);
        }
    }

    private static boolean isParsableMessage(String test, String pattern) {
        MessageFormat format = new MessageFormat(pattern);
        try {
            format.parse(test);
        } catch (ParseException e) {
            return false;
        }
        return true;
    }
}